### PR TITLE
qa: add filesystem/file sync stuck test support

### PR DESCRIPTION
**Pull request by lxbsz**
Issue1:

This will test the sync of the filesystem/file, which maybe stuck for
at most 5 seconds. This was because the related code will wait
for all the unsafe requests to get safe reply from MDSes, but the
MDSes just think that it's unnecessary to flush the mdlog immediately
after early reply, and the mdlog will be flushed every 5 seconds
in the tick thread.

This should have been fixed in kclient and libcephfs by triggering
mdlog flush before waiting requests' safe reply.

Issue2:

When doing the file/filesystem sync there is possible NULL pointer
dereference bug which will crash the kernel. Add a test case for this.

Fixes: https://tracker.ceph.com/issues/55329
Fixes: https://tracker.ceph.com/issues/55283
Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: qa/tasks/cephfs/test_misc.py [hunk 1]
```diff
@@ -231,6 +232,36 @@ def test_fs_lsflags(self):
         self.assertEqual(lsflags["allow_multimds_snaps"], True)
         self.assertEqual(lsflags["allow_standby_replay"], True)
 
+    def test_filesystem_sync_stuck_for_around_5s(self):
+        """
+        To check whether will the fsync will be be stuck to wait for the mdlog to be
```
Comment (kotreshhr) [inline qa/tasks/cephfs/test_misc.py]: nit: s/"To check whether will the fsync will be be stuck to wait"/"To check whether the fsync will be stuck waiting"/

File: qa/tasks/cephfs/test_misc.py [hunk 2]
```diff
@@ -231,6 +232,36 @@ def test_fs_lsflags(self):
...
+        flushed for at most 5 seconds.
+        """
+
+        dir_path = "fsync_do_not_wait_mdlog_testdir"
+        self.mount_a.run_shell(["mkdir", dir_path])
+
+        # run create/delete direcoties and test the sync time duration
```
Comment (kotreshhr) [inline qa/tasks/cephfs/test_misc.py:250]: nit: s/direcoties/directories/

File: qa/tasks/cephfs/test_misc.py [hunk 3]
```diff
@@ -232,35 +233,63 @@ def test_fs_lsflags(self):
...
-        """
-        To check whether will the fsync will be be stuck to wait for the mdlog to be
-        flushed for at most 5 seconds.
-        """
+    def _test_sync_stuck_for_around_5s(self, dir_path, file_sync=False):
 
-        dir_path = "fsync_do_not_wait_mdlog_testdir"
         self.mount_a.run_shell(["mkdir", dir_path])
 
+        pyscript = dedent("""
```
Comment (kotreshhr) [inline qa/tasks/cephfs/test_misc.py:241]: Could we rename `pyscript` to indicate sync directory, something like `sync_dir_pyscript` ?

File: qa/tasks/cephfs/test_misc.py [hunk 4]
```diff
@@ -232,35 +233,63 @@ def test_fs_lsflags(self):
...
             duration = time.time() - start
-            log.info(f"rmdir i = {i}, duration = {duration}")
+            log.info(f"sync rmdir i = {i}, duration = {duration}")
             self.assertLess(duration, 4)
 
         self.mount_a.run_shell(["rm", "-rf", dir_path])
 
+    def test_filesystem_sync_stuck_for_around_5s(self):
+        """
+        To check whether will the fsync will be be stuck to wait for the mdlog to be
```
Comment (kotreshhr) [inline qa/tasks/cephfs/test_misc.py:278]: nit: s/"To check whether will the fsync will be be stuck to wait"/"To check whether the fsync will be stuck waiting"/

File: qa/tasks/cephfs/test_misc.py [hunk 5]
```diff
@@ -232,35 +233,63 @@ def test_fs_lsflags(self):
...
+        flushed for at most 5 seconds.
+        """
+
+        dir_path = "filesystem_sync_do_not_wait_mdlog_testdir"
+        self._test_sync_stuck_for_around_5s(dir_path)
+
+    def test_file_sync_stuck_for_around_5s(self):
+        """
+        To check whether will the filesystem sync will be be stuck to wait for the
```
Comment (kotreshhr) [inline qa/tasks/cephfs/test_misc.py:287]: nit: s/"To check whether will the fsync will be be stuck to wait"/"To check whether the fsync will be stuck waiting"/

File: qa/tasks/cephfs/test_misc.py [hunk 6]
```diff
@@ -290,6 +291,71 @@ def test_file_sync_stuck_for_around_5s(self):
         dir_path = "file_sync_do_not_wait_mdlog_testdir"
         self._test_sync_stuck_for_around_5s(dir_path, True)
 
+    def test_file_filesystem_sync_crash(self):
+        """
+        To check whether will the kernel crash when doing the file/filesystem sync.
```
Comment (kotreshhr) [inline qa/tasks/cephfs/test_misc.py:296]: nit: s/"whether will"/"whether"/ ?

File: qa/tasks/cephfs/test_misc.py [hunk 7]
```diff
@@ -290,6 +291,71 @@ def test_file_sync_stuck_for_around_5s(self):
...
+
+        def mkdir_rmdir_thread(mount, path):
+            #global stop_thread
+
+            log.info(" mkdir_rmdir_thread starting...")
+            num = 0
+            while not stop_thread:
+                n = num
+                m = num
+                for i in range(10):
```
Comment (kotreshhr) [inline qa/tasks/cephfs/test_misc.py:311]: `for __ in range(10):` is preferred if `variable` is not used. https://docs.python-guide.org/writing/style/#create-an-ignored-variable

File: qa/tasks/cephfs/test_misc.py [hunk 8]
```diff
@@ -290,6 +291,71 @@ def test_file_sync_stuck_for_around_5s(self):
...
+                    mount.run_shell(["mkdir", os.path.join(path, f"{n}")])
+                    n += 1
+                for i in range(10):
```
Comment (kotreshhr) [inline qa/tasks/cephfs/test_misc.py:314]: Same as above.

File: qa/tasks/cephfs/test_misc.py [hunk 9]
```diff
@@ -290,6 +291,71 @@ def test_file_sync_stuck_for_around_5s(self):
...
+                    mount.run_shell(["rm", "-rf", os.path.join(path, f"{m}")])
+                    m += 1
+                num += 10
+            log.info(" mkdir_rmdir_thread stopped")
+
+        def filesystem_sync_thread(mount, path):
+            #global stop_thread
+
+            log.info(" filesystem_sync_thread starting...")
```
Comment (kotreshhr) [inline qa/tasks/cephfs/test_misc.py:323]: nit: space at the beginning intended ? Same comment for below instances.

**Comment by kotreshhr** [comment]
Can the commits 1da06da10b52227e99059b5d6f1821a0aa628ea0 and b271adea21bc30dabef3c9a8cde6343143470086 combined to make single commit ?

File: qa/tasks/cephfs/test_misc.py (hunk 7 above)
Comment (lxbsz) [reply qa/tasks/cephfs/test_misc.py:311]: Okay, will fix it.

File: qa/tasks/cephfs/test_misc.py (hunk 9 above)
Comment (lxbsz) [reply qa/tasks/cephfs/test_misc.py:323]: Yeah, that was intended.

**Review by lxbsz — COMMENTED** [review]
Thanks @kotreshhr

**Comment by lxbsz** [comment]
> Can the commits [1da06da](https://github.com/ceph/ceph/commit/1da06da10b52227e99059b5d6f1821a0aa628ea0) and [b271ade](https://github.com/ceph/ceph/commit/b271adea21bc30dabef3c9a8cde6343143470086) combined to make single commit ?

If not a must let's keep them in different commits to make it clearer what they are supposing to do ?

File: qa/tasks/cephfs/test_misc.py (hunk 7 above)
Comment (kotreshhr) [reply qa/tasks/cephfs/test_misc.py:314]: This is nit. Looks good to me.

**Comment by kotreshhr** [comment]
> > Can the commits [1da06da](https://github.com/ceph/ceph/commit/1da06da10b52227e99059b5d6f1821a0aa628ea0) and [b271ade](https://github.com/ceph/ceph/commit/b271adea21bc30dabef3c9a8cde6343143470086) combined to make single commit ?
> 
> If not a must let's keep them in different commits to make it clearer what they are supposing to do ?

Ack

**Review by kotreshhr — APPROVED** [review]

File: qa/tasks/cephfs/test_misc.py (hunk 7 above)
Comment (lxbsz) [reply qa/tasks/cephfs/test_misc.py:314]: Sorry, I thought I have fixed it. will push it again.

**Comment by lxbsz** [comment]
@kotreshhr 
Force pushed it by fixing the `for i in range(10):` issue, not sure I didn't update it that day.
Thanks

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2022-May-12

**Pull request merged**